common/src/main/java/com/proximyst/ban/service/MessageService.java

```diff
@@ -0,0 +1,230 @@
+      final @NonNull Punishment punishment) {
+    return this.messageFactory.placeholderMessage(messageKey,
+        this.messageFactory.componentComponent(this.createPlaceholders(punishment)));
+  }
+
+  private @NonNull CompletableFuture<@NonNull Void> announcePunishment(final @NonNull Punishment punishment,
+      final @NonNull MessageKey messageKey) {
+    final String permission = punishment.getPunishmentType().getNotificationPermission().orElse(null);
+    return this.punishmentMessage(messageKey, punishment)
+        .component()
+        .thenAccept(component -> {
+          this.banServer.consoleAudience().sendMessage(Identity.nil(), component);
+          for (final IBanAudience audience : this.banServer.onlineAudiences()) {
+            if (permission != null && !audience.hasPermission(permission)) {
+              continue;
+            }
+
+            audience.sendMessage(Identity.nil(), component);
+          }
+        });
+  }
+
+  private @NonNull CompletableFuture<@NonNull IMessageComponent @NonNull []> createPlaceholders(
+      final @NonNull Punishment punishment) {
+    return this.userService.getUser(punishment.getTarget())
+        .thenApply(
+            opt -> opt.orElseThrow(() -> new IllegalArgumentException("Target of punishment cannot be unknown.")))
+        .thenCombine(this.userService.getUser(punishment.getPunisher())
+                .thenApply(opt -> opt
+                    .orElseThrow(() -> new IllegalArgumentException("Punisher of punishment cannot be unknown."))),
+            (target, punisher) -> this.createPlaceholders(punishment, target, punisher))
+        .exceptionally(this.banExceptionalFutureLogger.cast());
+  }
+
+  private @NonNull IMessageComponent @NonNull [] createPlaceholders(final @NonNull Punishment punishment,
+      final @NonNull BanUser target, final @NonNull BanUser punisher) {
+    final IMessageComponent expiry =
+        !punishment.currentlyApplies()
+            ? this.messageFactory.keyComponent(KEY_PUNISHMENT_EXPIRY, MessageKey.FORMATTING_LIFTED)
+            : punishment.isPermanent()
+                ? this.messageFactory.keyComponent(KEY_PUNISHMENT_EXPIRY, MessageKey.FORMATTING_NEVER)
+                : this.messageFactory.staticComponent(KEY_PUNISHMENT_EXPIRY,
+                    DurationFormatUtils.formatDurationHMS(punishment.getExpiration() - System.currentTimeMillis()));
+    final IMessageComponent duration = punishment.isPermanent()
+        ? this.messageFactory.keyComponent(KEY_PUNISHMENT_DURATION, MessageKey.FORMATTING_PERMANENTLY)
+        : this.messageFactory
+            .staticComponent(KEY_PUNISHMENT_DURATION, MessageKey.FORMATTING_DURATION.map(this.messagesConfig)
+                .replace("<duration>",
+                    DurationFormatUtils.formatDurationWords(
+                        punishment.getExpiration() - System.currentTimeMillis(),
+                        false,
+                        false
+                    )));
```

Comment:
At least comment these monstrosities, please.

Reply:
I find it rather clear what's going on here. The indentation should be all the help you need :-)